ERCS/erc-7509.md

```diff
@@ -0,0 +1,329 @@
+title: Entity Component System
+description: Represent entities comprising components of data, with systems which operate on entities' components.
+author: Rickey (@HelloRickey)
+discussions-to: https://ethereum-magicians.org/t/a-new-proposal-of-entity-component-system/15665
+status: Draft
+type: Standards Track
+category: ERC
+created: 2023-09-05
+---
+
+## Abstract
+
+This proposal defines a minimal Entity Component System (ECS). Entities are unique identities that are assigned to multiple components (data) and then processed using the system (logic).
+This proposal standardizes the interface specification for using ECS in smart contracts, providing a set of basic functions that allow users to freely combine and manage multi-contract applications.
+
+## Motivation   
+
+ECS is a design pattern that improves code reusability by separating data from behavior. It is often used in game development. A minimal ECS consists of   
+**Entity**: a unique identifier.   
+**Component**: a reusable data container attached to an entity.   
+**System**: the logic for operating entity components.   
+**World**: a container for an entity component system.   
+This proposal uses smart contracts to implement an easy-to-use minimal ECS, eliminates unnecessary complexity, and makes some functional improvements that are consistent with contract interaction behavior. You can combine components and systems easily and freely.
+As a smart contract developer, the benefits of adopting ECS include:
+
+- It adopts a simple design of decoupling, encapsulation, and modularization, which makes the architecture design of your game or application easier.
+- It has flexible composition ability, each entity can combine different components. You can also define different systems for manipulating the data of these new entities.
+- It is conducive to expansion, and two games or applications can interact by defining new components and systems.
+- It can help your application add new features or upgrades, because data and behavior are separated, new features will not affect your old data.
+- It is easy to manage. When your application consists of multiple contracts, it will help you effectively manage the status of each contract.
+- Its components are reusable, and you can share your components with the community to help others improve development efficiency.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+World contracts are containers for entities, component contracts, and system contracts. Its core principle is to establish the relationship between entities and component contracts, and different entities will attach different components. And use the system contract to dynamically change the data of the entity in the component.
+Usual workflow when building ECS-based programs
```

Comment:
The wording and grammar are a bit off here. Does this get the correct point across?

```suggestion
World contracts are containers for entities, component contracts, and system contracts. Its core principle is to establish the relationship between entities and component contracts, where different entities will attach different components, and use system contracts to dynamically change the data of the entity in the component.

Usual workflow when building ECS-based programs:
```